### PR TITLE
types: expose `BeforeLoadFn` type

### PR DESCRIPTION
**Pull request by milhamm**
I encountered this issue when creating a function that needs the `BeforeLoadFn` types. I also noticed that other types such as `RouteLoaderFn` are exposed.

This PR is to expose the `BeforeLoadFn` type.

### CODE REVIEW

**Comment by SeanCassiere** [comment]
@schiller-manuel all good with exposing this type?

File: packages/react-router/src/route.ts [hunk 1]
```diff
@@ -205,7 +205,7 @@ export type BaseRouteOptions<
     getParentRoute: () => TParentRoute
   }
 
-type BeforeLoadFn<
+export type BeforeLoadFn<
```
[the remark this answers is not in the document]
Comment (milhamm) [reply packages/react-router/src/route.ts:208]: Got it, thanks

**Review by SeanCassiere — APPROVED** [review]

**Comment by SeanCassiere** [comment]
No objections from me about pushing this through.

@milhamm could you respond to Manuel's question? If he's fine with it, this can be pushed through.

**Comment by milhamm** [comment]
Sure @schiller-manuel @SeanCassiere.

So this is my example. In my case, It is not fully inferred, but I have a function that strictly handles only routes with search name with a typed context. Now the `beforeLoadWithNameSearch` has the typing for the search and also context.

```tsx
type WithNameSearch = {
  name?: string
}

const beforeLoadWithNameSearch: BeforeLoadFn<
  WithNameSearch,
  AnyRoute,
  any,
  RouteContext,
  MyRouterContext
> = ({ search, context }) => {
  console.log(search.name)
  console.log(context.user.name)
}

export interface MyRouterContext {
  user: {
    name?: string
  }
}

export const RootRoute = createRootRouteWithContext<MyRouterContext>()({
  component: () => {},
  notFoundComponent: () => {
    return <p>This is the notFoundComponent configured on root route</p>
  },
})

export const Route = createFileRoute('/posts/$postId/deep')({
  loader: async ({ params: { postId } }) => fetchPost(postId),
  errorComponent: PostErrorComponent as any,
  component: PostDeepComponent,
  validateSearch: (search: Record<string, unknown>): { name?: string } => {
    return {
      name: (search.name as string) || undefined,
    }
  },
  beforeLoad: beforeLoadWithNameSearch,
})
```

**Comment by schiller-manuel** [comment]
hm I don't think this type should be part of the public API, since we might want to modify the types / generics etc.
What do you really gain from using the `BeforeLoadFn` type? wouldn't this be much simpler?

```typescript
const beforeLoadWithNameSearch = (search : WithNameSearch, context : MyRouterContext) => {
  console.log(search.name)
  console.log(context.user.name)
}

export const Route = createFileRoute('/posts/$postId/deep')({
  loader: async ({ params: { postId } }) => fetchPost(postId),
  errorComponent: PostErrorComponent as any,
  component: PostDeepComponent,
  validateSearch: (search: Record<string, unknown>): { name?: string } => {
    return {
      name: (search.name as string) || undefined,
    }
  },
  beforeLoad: ({search, context}) => beforeLoadWithNameSearch(search, context),
})

**Comment by milhamm** [comment]
@schiller-manuel Yes you are right fair point, I don't see the benefit of exposing those to a public type now that you mentioned it. In fact, your example is the current implementation in my code right now.

**Pull request closed without merging**